packages/code-snippet/src/CodeSnippetWidget.tsx

```diff
@@ -214,37 +52,164 @@ export class CodeSnippetWidget extends ReactWidget {
+      namespace: METADATA_EDITOR_ID
+    });
+    this.restorer.restore(this.editorTracker, {
+      name: (metadataEditor: any) => {
+        return metadataEditor.metadata.name;
+      },
+      command: METADATA_EDITOR_ID + ':open',
+      args: (widget: MetadataEditor) => ({
+        name: widget.metadata.name,
+        description: widget.metadata.description,
+        language: widget.metadata.language,
+        code: widget.metadata.code,
+        newFile: widget.newFile
+      })
+    });
+    this.fetchData = this.fetchData.bind(this);
+    this.updateSnippets = this.updateSnippets.bind(this);
   }
 
   // Request code snippets from server
   async fetchData(): Promise<ICodeSnippet[]> {
     return await this.codeSnippetManager.findAll();
   }
 
-  // Triggered when the widget button on side palette is clicked
-  onAfterShow(msg: Message): void {
+  updateSnippets(): void {
     this.fetchData().then((codeSnippets: ICodeSnippet[]) => {
       this.renderCodeSnippetsSignal.emit(codeSnippets);
     });
   }
 
+  // Triggered when the widget button on side panel is clicked
+  onAfterShow(msg: Message): void {
+    this.updateSnippets();
+  }
+
+  addCodeSnippet(): void {
+    this.openCodeSnippetEditor('', '', '', '', true);
+  }
+
+  saveSnippet(snippetEditor: MetadataEditor): void {
+    const newSnippet = {
+      schema_name: 'code-snippet',
+      name: snippetEditor.metadata.name,
+      display_name: snippetEditor.metadata.name,
+      metadata: {
+        description: snippetEditor.metadata.description,
+        language: snippetEditor.metadata.language,
+        code: snippetEditor.editor.model.value.text.split('\n')
+      }
+    };
+    const newSnippetString = JSON.stringify(newSnippet);
+
+    if (snippetEditor.newFile) {
+      RequestHandler.makePostRequest(
+        snippetEditor.endpoint,
+        JSON.stringify(newSnippet),
+        false
+      ).then((response: any): void => {
+        this.updateSnippets();
+      });
+      snippetEditor.newFile = false;
+    } else {
+      RequestHandler.makeServerRequest(
+        snippetEditor.endpoint + '/' + newSnippet.name,
+        { method: 'PUT', body: newSnippetString },
+        false
+      ).then((response: any): void => {
+        this.updateSnippets();
+      });
+    }
+  }
+
+  async openCodeSnippetEditor(
+    displayName: string,
+    description: string,
+    language: string,
+    code: string,
+    newFile: boolean
+  ): Promise<void> {
+    const metadataEditorWidget = new MetadataEditor(
+      {
+        name: displayName,
+        description: description,
+        language: language,
+        code: code
+      },
+      newFile,
+      this.saveSnippet,
+      this.editorFactory,
+      CODE_SNIPPET_ENDPOINT
+    );
+    metadataEditorWidget.id = METADATA_EDITOR_ID;
+    if (newFile) {
+      metadataEditorWidget.title.label = 'New Snippet';
+    } else {
+      metadataEditorWidget.title.label = '[' + language + '] ' + displayName;
```

Comment:
this is a good candidate for template literals and avoiding having to have all the `+`:

```javascript
metadataEditorWidget.title.label = `$[{language}] ${displayName}`;
```